File: solutions/c/luhn/1/luhn.c
#include "luhn.h"
#include <ctype.h>


bool luhn(const char *num){
    unsigned int length = 0, digits = 0, m, sum = 0;
    while (num[length] != '\0'){
        if(isdigit(num[length])) digits++;
        length++;
    }
    if (digits <= 1) return false;
    for (int i = length - 1, j = 1; i >= 0; i--){
        if (num[i] == ' ') continue;
        if (isdigit(num[i])){
            m = num[i] - '0';
            if (j%2 == 0){
                m *= 2;
                if (m > 9) m -= 9;
            }
            sum += m;
            j++;
        }
        else return false;
    }
  return (sum%10) ? false : true;
}

File: solutions/julia/luhn/1/luhn.jl
function luhn(cardnum)
    num = replace(cardnum, ' ' => "")
    len = length(num)
    if len <= 1 || occursin(r"[^\d]", num); return false; end
    num = parse.(Int, collect(num))
    num[(len-1):-2:1] = (x -> 2x < 10 ? 2x : 2x - 9).(num[(len-1):-2:1])
    sum(num) % 10 == 0
end